google_authorized_keys/main.go

```diff
@@ -0,0 +1,226 @@
+			break
+		}
+		if i < 6 {
+			rt = time.Duration(3*i) * time.Second
+		}
+		logger.Errorf("error connecting to metadata server, retrying in %s, error: %v", rt, err)
+		time.Sleep(rt)
+	}
+	defer res.Body.Close()
+
+	md, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	return md, nil
+}
+
+func parseSSHKeys(username string, keys []string) []string {
+	var keyList []string
+	for _, key := range keys {
+		keySplit := strings.SplitN(key, ":", 2)
+		if len(keySplit) != 2 {
+			continue
+		}
+
+		user, keyVal, err := utils.GetUserKey(key)
+		if err != nil {
+			continue
+		}
+
+		if user == username {
+			keyList = append(keyList, keyVal)
+		}
+	}
+	return keyList
+}
+
+func getUserKeys(username string, instanceAttributes *attributes, projectAttributes *attributes) []string {
+	var userKeyList []string
+
+	instanceKeyList := parseSSHKeys(username, instanceAttributes.SSHKeys)
+	userKeyList = append(userKeyList, instanceKeyList...)
+
+	if !instanceAttributes.BlockProjectSSHKeys {
+
+		projectKeyList := parseSSHKeys(username, projectAttributes.SSHKeys)
+		userKeyList = append(userKeyList, projectKeyList...)
+
+	}
+
+	return userKeyList
+}
+
+func checkWinSSHEnabled(instanceAttributes *attributes, projectAttributes *attributes) bool {
+	if instanceAttributes.EnableWindowsSSH != nil {
+		return bool(*instanceAttributes.EnableWindowsSSH)
+	} else if projectAttributes.EnableWindowsSSH != nil {
+		return bool(*projectAttributes.EnableWindowsSSH)
+	}
+	return false
+}
+
+type attributes struct {
+	EnableWindowsSSH    *bool
+	BlockProjectSSHKeys bool
+	SSHKeys             []string
+}
+
+func getMetadataAttributes(metadataKey string) (*attributes, error) {
+	var a attributes
+	type jsonAttributes struct {
+		EnableWindowsSSH    string `json:"enable-windows-ssh"`
+		BlockProjectSSHKeys string `json:"block-project-ssh-keys"`
+		SSHKeys             string `json:"ssh-keys"`
+	}
+	var ja jsonAttributes
+	metadata, err := getMetadata(metadataKey, true)
+	if err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(metadata, &ja); err != nil {
+		return nil, err
+	}
+
+	value, err := strconv.ParseBool(ja.BlockProjectSSHKeys)
+	if err == nil {
+		a.BlockProjectSSHKeys = value
+	}
+
+	value, err = strconv.ParseBool(ja.EnableWindowsSSH)
+	if err == nil {
+		a.EnableWindowsSSH = &value
+	}
+	if ja.SSHKeys != "" {
+		a.SSHKeys = strings.Split(ja.SSHKeys, "\n")
+	}
+	return &a, nil
+}
+
+func main() {
+	ctx := context.Background()
+	username := os.Args[1]
+
+	opts := logger.LogOpts{
+		LoggerName:     programName,
+		FormatFunction: logFormat,
+	}
+
+	if runtime.GOOS == "windows" {
+		opts.Writers = []io.Writer{&serialPort{"COM1"}, os.Stderr}
+		opts.FormatFunction = logFormatWindows
+	} else {
+		opts.Writers = []io.Writer{os.Stderr}
+	}
+	logger.Init(ctx, opts)
+
+	instanceAttributes, err := getMetadataAttributes("instance/attributes/")
+	if err != nil {
+		logger.Errorf("Cannot read instance metadata attributes: %v", err)
```

Comment:
does this go to stdout or stderr? stdout is reserved for this binary, it should emit 0-or-more valid lines on stdout

Reply:
It goes to stderr, since the logger output is set to stderr above. All logger output in this script goes to stderr. The normal output is printed with fmt.Printf().